### PR TITLE
[WIP] Override list__action templates with config file

### PR DESCRIPTION
**Pull request by soullivaneuh**
PR related to #2586.
## Changelog

``` markdown
### Added
- `AbstractListBuilder` implementing `buildActionFieldDescription`
- List action templates can now be overriden with configuration file
```
## Subject

The goal is to permit `list__action.html.twig` and `list__action_*.html.twig` override from config file.
- [ ] Permit list templates options
- [ ] Update the documentation

@rande: I'm a little confusing. I passed list_action template on Configuration and compiler pass. How to pass it on `ListBuilder` classes?

Thanks.

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
Ok so I found a **huge** code duplicate:

```php
/**
 * @param FieldDescriptionInterface $fieldDescription
 *
 * @return FieldDescriptionInterface
 */
public function buildActionFieldDescription(FieldDescriptionInterface $fieldDescription)
{
    if (null === $fieldDescription->getTemplate()) {
        $fieldDescription->setTemplate('SonataAdminBundle:CRUD:list__action.html.twig');
    }

    if (null === $fieldDescription->getType()) {
        $fieldDescription->setType('action');
    }

    if (null === $fieldDescription->getOption('name')) {
        $fieldDescription->setOption('name', 'Action');
    }

    if (null === $fieldDescription->getOption('code')) {
        $fieldDescription->setOption('code', 'Action');
    }

    if (null !== $fieldDescription->getOption('actions')) {
        $actions = $fieldDescription->getOption('actions');
        foreach ($actions as $k => $action) {
            if (!isset($action['template'])) {
                $actions[$k]['template'] = sprintf('SonataAdminBundle:CRUD:list__action_%s.html.twig', $k);
            }
        }

        $fieldDescription->setOption('actions', $actions);
    }

    return $fieldDescription;
}
```

Present on each persistence layer bundle.

I'll create an abstract list builder and move the code on this part.

Then, admin dependency will be bumped on each bundle.

**Comment by Soullivaneuh** [comment]
Config template seems to be injected to the admin service, but I can't access it from the builder...

I don't really know how to do it properly. Any idea @sonata-project/contributors?

cc @greg0ire

File: Tests/Fixtures/Builder/FakeListBuilder.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+use Sonata\AdminBundle\Admin\AdminInterface;
+use Sonata\AdminBundle\Admin\FieldDescriptionCollection;
+use Sonata\AdminBundle\Admin\FieldDescriptionInterface;
+use Sonata\AdminBundle\Builder\AbstractListBuilder;
+
+/**
+ * @author Sullivan Senechal <soullivaneuh@gmail.com>
+ */
+final class FakeListBuilder extends AbstractListBuilder
```
Comment (core23) [inline Tests/Fixtures/Builder/FakeListBuilder.php]: You can remove the final key

File: Tests/Fixtures/Builder/FakeListBuilder.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function fixFieldDescription(AdminInterface $admin, FieldDescriptionInterface $fieldDescription)
+    {
+        // TODO: Implement fixFieldDescription() method.
```
Comment (core23) [inline Tests/Fixtures/Builder/FakeListBuilder.php]: Please remove the TODOs

**Comment by core23** [comment]
Ping @Soullivaneuh

File: Tests/Fixtures/Builder/FakeListBuilder.php (hunk 1 above)
Comment (Soullivaneuh) [reply Tests/Fixtures/Builder/FakeListBuilder.php]: Why? This class should not be extended.

File: Tests/Builder/AbstractListBuilderTest.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace Sonata\AdminBundle\Tests\Builder;
+
+use Sonata\AdminBundle\Tests\Fixtures\Admin\FieldDescription;
+use Sonata\AdminBundle\Tests\Fixtures\Builder\FakeListBuilder;
+
+/**
+ * @author Sullivan Senechal <soullivaneuh@gmail.com>
+ */
+class AbstractListBuilderTest extends \PHPUnit_Framework_TestCase
```
Comment (Soullivaneuh) [inline Tests/Builder/AbstractListBuilderTest.php]: Missing `final`

**Comment by Soullivaneuh** [comment]
@core23 At the last state of this PR, I asked for some advice.

I also activated edition from maintainers if somebody want to update this PR.

If anybody also want to continue, feel free to notify and open a new PR.

File: DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 4]
```diff
@@ -375,6 +375,10 @@ public function fixTemplates(ContainerBuilder $container, Definition $definition
             'ajax' => 'SonataAdminBundle::ajax_layout.html.twig',
             'dashboard' => 'SonataAdminBundle:Core:dashboard.html.twig',
             'list' => 'SonataAdminBundle:CRUD:list.html.twig',
+            'list__action' => 'SonataAdminBundle:CRUD:list__action.html.twig',
```
Comment (core23) [inline DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Do we really need this override? You can replace this with the normal twig template overrides using `ROOT/templates/SonataAdminBundle/...`?

**Pull request closed without merging**